File: ldap-poc/src/main/java/org/example/ldappoc/controller/UserController.java
package org.example.ldappoc.controller;

import org.example.ldappoc.model.LdapUser;
import org.example.ldappoc.service.LdapGroupService;
import org.example.ldappoc.service.LdapUserService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/users")
public class UserController {

    private final LdapUserService userService;
    private final LdapGroupService groupService;

    public UserController(LdapUserService userService, LdapGroupService groupService) {
        this.userService = userService;
        this.groupService = groupService;
    }

    @GetMapping
    public List<LdapUser> findAll() {
        return userService.findAll();
    }

    @GetMapping("/{uid}")
    public ResponseEntity<LdapUser> findByUid(@PathVariable String uid) {
        return userService.findByUid(uid)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/search")
    public List<LdapUser> search(@RequestParam String name) {
        return userService.searchByName(name);
    }

    @GetMapping("/by-title")
    public List<LdapUser> searchByTitle(@RequestParam String title) {
        return userService.searchByTitle(title);
    }

    @GetMapping("/{uid}/groups")
    public List<String> getUserGroups(@PathVariable String uid) {
        return groupService.findGroupsForUser(uid);
    }

    @PostMapping("/authenticate")
    public ResponseEntity<Map<String, Object>> authenticate(@RequestBody Map<String, String> credentials) {
        String uid = credentials.get("uid");
        String password = credentials.get("password");
        boolean authenticated = userService.authenticate(uid, password);
        return ResponseEntity.ok(Map.of(
                "uid", uid,
                "authenticated", authenticated
        ));
    }
}

File: kotlin-value-class-poc/src/main/kotlin/org/example/valueclasspoc/controller/UserController.kt
package org.example.valueclasspoc.controller

import org.example.valueclasspoc.model.Email
import org.example.valueclasspoc.model.Money
import org.example.valueclasspoc.model.User
import org.example.valueclasspoc.model.UserId
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/users")
class UserController {

    @GetMapping
    fun getUsers(): List<User> = listOf(
        User(UserId(1), "Alice", Email("alice@example.com"), Money(10050)),
        User(UserId(2), "Bob", Email("bob@example.com"), Money(25099))
    )

    @PostMapping("/validate")
    fun validateEmail(@RequestBody body: Map<String, String>): ResponseEntity<Map<String, Any>> {
        val emailStr = body["email"] ?: return ResponseEntity.badRequest()
            .body(mapOf("valid" to false, "error" to "email field is required"))
        return try {
            Email(emailStr)
            ResponseEntity.ok(mapOf("valid" to true, "email" to emailStr))
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("valid" to false, "error" to e.message.orEmpty()))
        }
    }
}
